### ISSUE TITLE
[3.0.1] After interruption, unable to launch the App from device home screen on first attempt

### ISSUE THREAD

**Issue by Sdaswani** [mozilla-mobile/firefox-tv#1257]
I'm getting more info from @haldemann99 on this.

**Comment by Sdaswani** [comment]
@haldeman99 is checking but he believes this happens after #1256.

**Comment by mcomella** [comment]
I just reproduced this and seem to have received platform crashes:
```
2018-10-08 14:31:56.547 29865-29995/? E/cr_MediaResource: Cannot parse uri: java.lang.IllegalArgumentException: Expected scheme-specific part at index 5: blob:
    java.lang.IllegalArgumentException: Expected scheme-specific part at index 5: blob:
        at java.net.URI.create(URI.java:853)
        at org.chromium.content.browser.MediaResourceGetter.configure(MediaResourceGetter.java:208)
        at org.chromium.content.browser.MediaResourceGetter.extractMediaMetadata(MediaResourceGetter.java:1139)
     Caused by: java.net.URISyntaxException: Expected scheme-specific part at index 5: blob:
        at java.net.URI$Parser.fail(URI.java:2855)
        at java.net.URI$Parser.failExpecting(URI.java:2861)
        at java.net.URI$Parser.parse(URI.java:3064)
        at java.net.URI.<init>(URI.java:589)
        at java.net.URI.create(URI.java:851)
        at org.chromium.content.browser.MediaResourceGetter.configure(MediaResourceGetter.java:208) 
        at org.chromium.content.browser.MediaResourceGetter.extractMediaMetadata(MediaResourceGetter.java:1139) 
```

and:
```
2018-10-08 14:32:00.313 29865-29865/? E/cr_Cloud9JNICrash: java.lang.NullPointerException: Attempt to invoke virtual method 'android.view.SurfaceHolder android.view.SurfaceView.getHolder()' on a null object reference
        at org.chromium.amazon_webview.browser.media.AmazonMediaPlayerBridge.setSurfaceView(AmazonMediaPlayerBridge.java:260)
        at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method)
        at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:41)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6188)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:911)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:801)
```

As well as a native thread dump. Here are the full logs: [i257-logs.txt.tar.gz](https://github.com/mozilla-mobile/firefox-tv/files/2458059/i257-logs.txt.tar.gz)

I'll try again to verify these are the correct traces.

**Comment by mcomella** [comment]
This reproduce on a v2.2 LAT, fwiw.

**Comment by mcomella** [comment]
I got the same error, note that the dump happened when hitting home, not when I opened the app.

It's possible we're doing something wonky with the API that is causing this (e.g. we call `onPause` from `onStop`). Also, I saw `onResume` in the logs but we're stopping: 
```
2018-10-08 14:39:51.161 30914-30914/? E/com.amazon.android.webkit.android.AndroidWebViewDelegate: Method not supported in Android WebView
    java.lang.UnsupportedOperationException
        at com.amazon.android.webkit.android.AndroidWebViewDelegate.getWebViewReflection(AndroidWebViewDelegate.java:901)
        at com.amazon.android.webkit.android.AndroidWebViewDelegate.onResume(AndroidWebViewDelegate.java:383)
        at com.amazon.android.webkit.AmazonWebView.onResume(AmazonWebView.java:826)
        at org.mozilla.focus.webview.FirefoxAmazonWebView.onStart(FirefoxAmazonWebView.kt:62)
        at org.mozilla.focus.iwebview.IWebViewLifecycleFragment.onStart(IWebViewLifecycleFragment.kt:80)
        at android.support.v4.app.Fragment.performStart(Fragment.java:2380)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1458)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1740)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1809)
        at android.support.v4.app.BackStackRecord.executeOps(BackStackRecord.java:799)
        at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2580)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2367)
        at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2322)
        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2229)
        at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:700)
        at android.os.Handler.handleCallback(Handler.java:751)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6188)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:911)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:801)
```

**Comment by mcomella** [comment]
Okay, `onResume` and first exception occur when starting the WebView and clicking a video respectively. The crash, which occurs when home is pressed, has the following stack trace:
```
2018-10-08 14:32:00.313 29865-29865/? E/cr_Cloud9JNICrash: java.lang.NullPointerException: Attempt to invoke virtual method 'android.view.SurfaceHolder android.view.SurfaceView.getHolder()' on a null object reference
        at org.chromium.amazon_webview.browser.media.AmazonMediaPlayerBridge.setSurfaceView(AmazonMediaPlayerBridge.java:260)
        at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method)
        at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:41)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6188)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:911)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:801)
```

And the native dump from https://github.com/mozilla-mobile/firefox-tv/issues/1257#issuecomment-427987516.

I should try to reproduce on the latest LAT, I wonder if this is related to the WebViewCache.

**Comment by mcomella** [comment]
I reproduced this on v2.2 from the app store which is not using our WebView cache, invalidating my concern above.

We could still be doing something wonky.

**Comment by mcomella** [comment]
I don't see any crash reports in Sentry: I'm guessing it doesn't appear because it's a native crash (which maybe logs a related Java Exception?).

Since we haven't seen this before, I'm tempted to say it's a platform crash introduced by a new version of WebView but I don't actually know if the version of WebView has changed.

I'll try to reproduce on a debug build and add some logging to get a better sense of what's really happening.

**Comment by mcomella** [comment]
I can't reproduce this on a local debug or release build on the `releases/v3.0` branch using a different app ID: perhaps WebView is taking different behavior based on our package ID.

**Comment by mcomella** [comment]
I can reproduce from `releases/v3.0` when using the release app ID: seems questionable.

**Comment by mcomella** [comment]
This problem does not occur when I moved `WebView.onPause` (and onResume) from `onStop` to `onPause`. However, we set up the code like that to work around another issue with MediaSession and I confirmed this problem is still reproducible.

### Issue summary
STR:
- Open Firefox on the release app ID (org.mozilla.tv.firefox; doesn't matter if it's a debug or release build)
- Open youtube
- Open a video
- Click home

Expected: Goes home, no crash, and reopening resumes the video
Actual: Native crash! Sometimes you need to click on Firefox twice to resume and it will resume from the home screen. See See trace in https://github.com/mozilla-mobile/firefox-tv/issues/1257#issuecomment-427989646

This problem does not occur when `webView.onPause` is moved from `EngineViewLifecycleFragment.onStop` to `EngineViewLifecycleFragment.onPause` (same with onResume). This is a workaround to #936. From our end, there is nothing we can do (unless we want to break MediaSession with voice commands).

@Sdaswani Can you ask Amazon to look into this bug? Could you also ask if different code is being executed for our release app ID? This problem doesn't occur on app IDs other than our release so I suspect it's happening which is undesirable because 1) we have no guarantees about what this code is doing and 2) sometimes it creates inconsistencies like this one. Can you ask about that as well?

**Comment by Sdaswani** [comment]
@haldeman99 this seems to be something on the fire os side - can you please have your team take a look? https://github.com/mozilla-mobile/firefox-tv/issues/1257#issuecomment-428377953

**Comment by Sdaswani** [comment]
@mcomella @haldeman99 is checking if his QA team can repro this with the netflix STRs. If so, he will ask for a waiver of it as an RB.

**Comment by mcomella** [comment]
> This problem does not occur when webView.onPause is moved from EngineViewLifecycleFragment.onStop to EngineViewLifecycleFragment.onPause (same with onResume). This is a workaround to #936. From our end, there is nothing we can do (unless we want to break MediaSession with voice commands).

To elaborate on why we needed this workaround, when `WebView.onPause` is moved to `onPause` and similarly for resume:
STR
- On a play youtube TV video
- User says "Alexa, pause"

Expected: Video pauses for voice command and remains paused
Actual: Video goes to black screen for voice command, starts to buffer (sometimes reloading the video), and continues playing once the voice command ends

iirc (and it's been a while), there are a few problems here:
- The screen goes black: not sure why. Maybe `onPause/pauseTimers` forces the screen to blank?
- The video rebuffers: not sure why. Maybe `onPause/pauseTimers`? From past experience, I think YouTube TV is especially prone to rebuffering from interruption
- The video doesn't pause after the voice command: WebView is managing its own audio focus so we don't receive any callbacks when the voice command completes / the WebView resumes. As such, we don't know when we can send the MediaSession play/pause event (since we can't execute the javascript to pause the WebView's video element until the WebView is running).

All of these issues were solved by moving the callbacks to `onStop`. iirc, we're able to detect when the voice command ends by detecting the onPlay/onPause events in JavaScript and queuing any MediaSession commands until the video starts playing again (hence the slight delay between play/pause). I forget why this didn't work with `onStop/onPause`.

This issue is a combination of a few APIs that aren't playing well together:
- MediaSession
- WebView (and the way it manages audio focus so we can't see MediaSession events)
- Firefox injecting JavaScript to manage MediaSession state (and the WebView not always being available)

**Comment by liuche** [comment]
I'll also investigate further, to see if there's a hack while we wait for a fix from the Amazon side.

**Comment by liuche** [comment]
Okay, I've done some more testing and I have a question for @haldeman99:

It looks like there's some Fire OS process that is listening for the Firefox app id, and trying to save the SurfaceView state during `onPause` - can you explain a little what that process is so we can call into that directly? This process doesn't save the state fast enough so it causes our app to crash because the SurfaceView is destroyed by the time it runs.

For more context:
I think the Fire OS process (`ExoPlayerStateMachineImplBase: setSurface API call. Block= true`? `ExternalPlaybackStateManager`?) is trying to save the state of the SurfaceView (that's enforced for our Firefox app id - this doesn't happen for our debug app id) but since we call `onPause` so late this results in a crash because the SurfaceView is destroyed before this can happen, and then it fails to launch from the "recents".

The reason we call`onPause` so late is a workaround for how we implemented MediaSession "pause" commands - since the Alexa voice command pauses the app, we have to handle the "Alexa, pause" command in the `onStart` method, otherwise it gets lost. Ideally we could get AudioFocus changes but I think @mcomella wasn't able to get that API.

So there are two options here:
a) If we can hear from @haldeman99 what the command to save SurfaceView state is, we can just call that directly in onPause
b) We can break the "Alexa, pause" command, but everything else will work - "play", "fast forward", "next video" etc.

**Comment by liuche** [comment]
@Sdaswani I'm going to close this bc this should be fixed in OS update.
